### PR TITLE
HS-1627: add certificate revocation when delete location

### PR DESCRIPTION
**Pull request by freddy61025**
## Description
Align with PKI add revoke certificate and validate certificate serial number in certificate manger
BFF delete location will trigger revoke certificate if location successfully removed 

Will followup with https://opennms.atlassian.net/browse/HS-1737

## Jira link(s)
- https://opennms.atlassian.net/browse/HS-1626
- https://opennms.atlassian.net/browse/HS-1627

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts

### CODE REVIEW

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ */
+@Component
+public class SerialNumberRepository {
+    private final ObjectMapper mapper = new ObjectMapper();
+    private static final Logger LOG = LoggerFactory.getLogger(SerialNumberRepository.class);
+
+    private final RocksDB db;
+    public SerialNumberRepository(@Value("${grpc.server.db:/app/db}") String rootDir) throws RocksDBException{
+        Objects.requireNonNull(rootDir);
+        LOG.info("Begin to init rocksDB of {}", rootDir);
```
Comment (artnaseef) [inline minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java:60]: Grammar nit.  Maybe reword to "Beginning init of rocksDB at {}".

Another thought - I personally use the logging syntax "<message>: <key>=<value>; ..." - goes back to using Splunk.  So in this case, I would personally do something like this - `Beginning init of rocksDB: db-path={}`

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        LOG.info("Begin to close rocketDb");
+        if (db != null) {
+            db.close();
+        }
+        LOG.info("Successfully closed rocketDb");
+    }
+
+    public void addCertificate(String tenantId, String locationId, X509Certificate certificate) throws RocksDBException, IOException {
+        var meta = new CertificateMeta(tenantId, locationId, certificate);
+        db.put(meta.getSerial().getBytes(), mapper.writeValueAsBytes(meta));
```
Comment (artnaseef) [inline minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java:83]: Can the key use the tenant-id and location-id so lookups avoid the linear search?  If rocks DB doesn't support a composite key, then we could format something like `TENANT-ID=<tenant>|LOCATION=<location>` - a syntax like that should make conflicting keys almost impossible, and completely infeasible.

Another question - is it possible in rocks DB to have the value be a list?  If not, perhaps we can store the value as JSON text containing a list of the valid certificate serial numbers?

Looking ahead, I see there's a need to be able to lookup the cert just by serial number... I'm going to reach out offline to discuss more.

**Review by artnaseef — COMMENTED** [review]
Didn't finish; want to catch up offline to discuss...

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java (hunk 2 above)
Comment (freddy61025) [reply minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java:83]: The main target is to make certificate don't contain any custom values in subject. We only use serial number as key.

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+/**
+ * DB structure is key (serial number) value (metadata)
+ */
+@Component
+public class SerialNumberRepository {
+    private final ObjectMapper mapper = new ObjectMapper();
+    private static final Logger LOG = LoggerFactory.getLogger(SerialNumberRepository.class);
+
+    private final RocksDB db;
+    public SerialNumberRepository(@Value("${grpc.server.db:/app/db}") String rootDir) throws RocksDBException{
```
[the remark this answers is not in the document]
Comment (freddy61025) [reply minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java:58]: I will change it. It just my personal preference to make names more generic. That switching backend will not affect paths.

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java (hunk 2 above)
Comment (artnaseef) [reply minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java:83]: On follow-up, we don't need to optimize this solution, so no changes needed here.

File: minion-certificate-manager/main/src/test/resources/test-output-dir/client.signed.cert
Comment (artnaseef) [inline minion-certificate-manager/main/src/test/resources/test-output-dir/client.signed.cert:1]: Is this file a test input or test output?

If output, then please delete.  I recommend also adding the folder to the `.gitignore` file.  Or, better yet, move the test path to use the `target/` folder instead of `src/test/resources` for test output - tests should never write to any place under `src/`.

File: rest-server/src/main/java/org/opennms/horizon/server/service/GrpcLocationService.java [hunk 4]
```diff
@@ -86,6 +88,11 @@ public Mono<MonitoringLocation> updateLocation(@GraphQLArgument(name = "location
 
     @GraphQLMutation
     public Mono<Boolean> deleteLocation(@GraphQLArgument(name = "id") long id, @GraphQLEnvironment ResolutionEnvironment env) {
-        return Mono.just(client.deleteLocation(id, headerUtil.getAuthHeader(env)));
+        var accessToken = headerUtil.getAuthHeader(env);
+        var status = client.deleteLocation(id, accessToken);
+        if (status) {
```
Comment (artnaseef) [inline rest-server/src/main/java/org/opennms/horizon/server/service/GrpcLocationService.java:93]: It might be good to call revoke even if delete fails, because it is possible the location could get deleted but the certificate doesn't get revoked due to a service crash or other scenario.

**Review by artnaseef — CHANGES_REQUESTED** [review]

File: minion-certificate-manager/main/src/test/resources/test-output-dir/client.signed.cert
Comment (freddy61025) [reply minion-certificate-manager/main/src/test/resources/test-output-dir/client.signed.cert:1]: It is a test input. It is for PKCS12GeneratorTest.
PKCS12GeneratorTest itself is mocking the CommandExecutor, openssl is never be called. That said getX509Certificate will never able to find the real cert. The only way i can think of is put a fake file in test output folder. The other way is to spy PKCS12Generator and override getX509Certificate. (However for PKCS12GeneratorTest, I don't want to spy PKCS12Generator)

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (freddy61025) [reply minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java]: done

File: rest-server/src/main/java/org/opennms/horizon/server/service/GrpcLocationService.java (hunk 4 above)
Comment (freddy61025) [reply rest-server/src/main/java/org/opennms/horizon/server/service/GrpcLocationService.java]: done

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java (hunk 1 above)
Comment (freddy61025) [reply minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/SerialNumberRepository.java]: done

File: minion-certificate-manager/docker-it/src/test/java/org/opennms/horizon/minioncertmanager/stepdefs/MinionCertificateManagerTestSteps.java [hunk 5]
```diff
@@ -85,36 +97,65 @@ public void createGrpcConnection() {
...
             100,
             timeout,
-            false);
-        assertTrue("Zip file created", success);
+            null);
+        this.serialNumber = serialNumber;
+        assertNotNull("P12 file created", serialNumber);
+    }
+
+    @Then("send isValid with last serial number and timeout {int}ms")
```
Comment (artnaseef) [inline minion-certificate-manager/docker-it/src/test/java/org/opennms/horizon/minioncertmanager/stepdefs/MinionCertificateManagerTestSteps.java:123]: The wording on this step is a little unclear to me.  How about the following instead?

    send "isValid" request with last serial number and timeout {int}ms

File: minion-certificate-manager/docker-it/src/test/java/org/opennms/horizon/minioncertmanager/stepdefs/MinionCertificateManagerTestSteps.java [hunk 6]
```diff
@@ -85,36 +97,65 @@ public void createGrpcConnection() {
...
+    public String checkLastIsValid(long timeout) throws InterruptedException {
+        LOG.info("Checking certificate serial number: {}", serialNumber);
+        Supplier<IsCertificateValidResponse> call = () -> clientUtils.getMinionCertificateManagerStub()
+            .isCertValid(IsCertificateValidRequest.newBuilder().setSerialNumber(serialNumber).build());
+        Supplier<IsCertificateValidResponse> isValid = retryUtils.retry(
+            () -> call,
+            result -> result.get().getIsValid(),
+            100,
+            timeout,
+            null);
```
Comment (artnaseef) [inline minion-certificate-manager/docker-it/src/test/java/org/opennms/horizon/minioncertmanager/stepdefs/MinionCertificateManagerTestSteps.java:133]: Is there some timing here where it might come back as invalid but then later come back as valid?  If not, then the retries can be removed.

File: minion-certificate-manager/main/src/test/java/org/opennms/horizon/minioncertmanager/certificate/PKCS12GeneratorTest.java [hunk 7]
```diff
@@ -71,6 +74,8 @@ void testGenerateP12() throws IOException, InterruptedException {
         // Verify the Results
         //
         Mockito.verify(mockCommandExecutor).executeCommand("openssl genrsa -out client.key.pkcs1 2048", outputDirPath.toFile());
+        Mockito.verify(mockCertificateReader).getX509Certificate(
+            outputDirPath.toFile().getAbsolutePath() + FileSystems.getDefault().getSeparator() + "client.signed.cert");
```
Comment (artnaseef) [inline minion-certificate-manager/main/src/test/java/org/opennms/horizon/minioncertmanager/certificate/PKCS12GeneratorTest.java:78]: I like the use of the `mockCertificateReader`.

One thing I recommend here - instead of directly verifying the `getX509Certificate` method was called, use a `Mockito.when(...)` on that call and have it return a value.  Then, capture the return value fro the call to `generate` above, and compare them.

The advantage of that approach is not only does it make sure the right method call is made, it also ensures the result of that method call is returned by the `generate` method itself.  Then the `Mockito.verify(...)` of that method call is no longer needed, because the test already proves it is called.

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/CertificateReader.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+ *******************************************************************************/
+package org.opennms.horizon.minioncertmanager.certificate;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+
+public class CertificateReader {
```
Comment (artnaseef) [inline minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/certificate/CertificateReader.java:36]: This looks good.

**Review by artnaseef — CHANGES_REQUESTED** [review]

File: minion-certificate-manager/docker-it/src/test/java/org/opennms/horizon/minioncertmanager/stepdefs/MinionCertificateManagerTestSteps.java (hunk 6 above)
Comment (freddy61025) [reply minion-certificate-manager/docker-it/src/test/java/org/opennms/horizon/minioncertmanager/stepdefs/MinionCertificateManagerTestSteps.java]: i just think using the retry util is a common practices.

File: minion-certificate-manager/main/src/test/java/org/opennms/horizon/minioncertmanager/certificate/PKCS12GeneratorTest.java (hunk 7 above)
Comment (freddy61025) [reply minion-certificate-manager/main/src/test/java/org/opennms/horizon/minioncertmanager/certificate/PKCS12GeneratorTest.java]: done

**Review by artnaseef — APPROVED** [review]
